19.0 estate stfra

stfra-odoo opened the pull request.

stfra-odoo committed:
[ADD] estate:  Create a new module for real estate advertisment with a model set

stfra-odoo committed:
[FIX] estate: Fix required field needed in expected price and a typo

robodoo commented:
[![Pull request status dashboard](https://mergebot.odoo.com/odoo/tutorials/pull/1075.png)](https://mergebot.odoo.com/odoo/tutorials/pull/1075)

stfra-odoo committed:
[ADD] estate:  Add some access rights to the db

stfra-odoo committed:
[IMP] estate:  enable UI interaction trough menus, display property data in list view and create/edit property data in form view

The model was created and access rights provided but there was no interaction with the model.  This is implemented through xml as its is mre convinient for complex data. There is one menu to acces the list view of the properties and a form views is displayed whend creating or editing a property. Some fields have now a default value, some are read-only, some are requiered and some cannot be copied when duplicated. The fields active and state were added for properties display in the list view by default and for it to have a different state depending on the stage respectively.

stfra-odoo committed:
[IMP] estate: enrich the different views  (list, form, and search) of the estate module.

The views created were basic one and need some custumization to be use in a business application.

List view:
the following fields have been added: Title, Postcode, bedrooms, Living Area (sqm), Expected Price, Selling Price, Available From

Form view:
It is now composed of groups and a notebook with a page description

Search view:
Search has now other predifened shorcuts (Title, Postcode, expectedPrice, Bedrooms, Living Area, Facades), a predefined filter "Available" (that looks for properties with the state New or Offer received) and a Group by filter on the postcodes

stfra-odoo committed:
[IMP] estate: introduce various new data in the estate property model: buyer, salesperson, property type, list of tags and list of offers.

Buyer
Can be any individual (res.partner)

Salesperson
Has to be an employe (res.user)

Property type
New model with the name of the type (estate.property.type)

Tags
New model with the name of the tag (estate.property.tag)

Offers
New model with the buyer, price, status and property id (estate.property.offer)

All those new models have list and form view but the offer model is only accessible from the porperty form (as each offer is linked to a property)

A new menu setting has been added to allow the creation of new tags and types.

ushyme commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,41 @@
+from datetime import timedelta
+
+def default_availability_date(recordset):
+    return fields.Datetime.today() + timedelta(days=90)
```
```suggestion
    return fields.Date.context_today(self) + timedelta(days=90)
```
It's better to use `context_today()` here so that it takes the user's timezone into account.

ushyme reviewed (commented):
Good work so far; I just have one small remark.

stfra-odoo committed:
[IMP] estate: enrich the models with new fields (best price and total area in estate property and validity and deadline in property offer).

Best price is computed on the maximum price of all the offers linked to the property.

Total area is computed as the sum of the lving area and garden area.

Validity is the number days that have passed since the creation of the offer during wich it remains valid.

Deadline is the date until wich the offer remains valid. It is computed based on validity and when changed, updates validity.

stfra-odoo committed:
[LINT] estate: format file of estate module to remove linting issues.

Previewsly made pull request could not pass because of the linting issues.

ushyme commented:
Runbot is still red 😃

stfra-odoo committed:
[IMP] estate: allow more direct interactions with the property and its offers.

The state of a property and its offer could only be changed manually which is not safe way of handling those status.

Now buttons have been added : sold and cancel for the property and accept and refuse for the offers. An error pop-up is shown when trying to perfom an illegal action.

stfra-odoo committed:
[LINT] estate: add missing string to button with icon in offer view xml.

stfra-odoo committed:
[IMP] estate: check data consistency using SQL and python constraints on the models.

Estate property model:
- SQL constraints: check that theexpected price is stricly positive and selling price positive.
- Python constraint: check that the selling price is at least 90% of the expected price when one of them is modified.

Estate property offer model:
- SQL constraint: check that the price is strictly positive.

Estate property tag model:
- SQL constraint: check that the name tag is unique.

Estate property type model:
- SQL constraint: check that the name type is unique.

stfra-odoo committed:
[IMP] estate: imporove user interface with some business logic and buttons.

State are now displayed in a status bar. The buttons (Sold/cancel on properties and Accept/Refuse on offers) are now dynamic and only appeares when needed. Offers can be created and edited in place. Tags have colors. Property type can only be created and edited from the settings. Offers related to a  property type can be acces from its form.

ushyme commented on estate/models/estate_property_offer.py:
```diff
@@ -0,0 +1,57 @@
+        for record in self:
+            record.status = 'refused'
+            record.property_id.property_buyer_id = None
+            record.property_id.selling_price = 0
```
This currently clears the property's selling price even if you are refusing a random low offer. It should only clear the property fields if the currently accepted offer is the one being refused.
```suggestion
    def action_refuse(self):
        for record in self:
            if record.status == 'accepted':
                record.property_id.selling_price = 0
                record.property_id.property_buyer_id = False
            record.status = 'refused'
```

ushyme commented on estate/security/ir.model.access.csv:
```diff
@@ -0,0 +1,5 @@
+access_real_estate_platform,estate_property,model_estate_property,base.group_user,1,1,1,1
+access_real_estate_type,estate_property_type,model_estate_property_type,base.group_user,1,1,1,1
+access_real_estate_tag,estate_property_tag,model_estate_property_tag,base.group_user,1,1,1,1
+access_real_estate_offer,estate_property_offer,model_estate_property_offer,base.group_user,1,1,1,1
```
```suggestion
access_real_estate_offer,estate_property_offer,model_estate_property_offer,base.group_user,1,1,1,1

```
missing new line at the end of the file.
this goes for most of your other non-python files as well

ushyme commented on estate/views/estate_property_views.xml:
```diff
@@ -0,0 +1,114 @@
+        <field name="arch" type="xml">
+            <form string="Estate Property">
+                <header>
+                    <button name="action_sold" type="object" string="Sold" invisible="state in ['new', 'offer_received', 'sold', 'cancelled']" class="btn btn-primary"/>
```
```suggestion
                    <button name="action_sold" type="object" string="Sold" invisible="state != 'offer_accepted'" class="btn btn-primary"/>
```

stfra-odoo committed:
[IMP] estate: alter the standart delete method of properties, the standart creation method of offers and the view form of an user.

The new delete method with the decorator 'ondelete()' prevent the deletion of a property that is not in the state 'New' or 'Cancelled'.

The new creation method prevent user from creatin an offer with a lower amount than an existing offer and change the state of the property to 'Offer recieved' when it was 'New'.

The list of available properties is now displayed in a new page on the user form view.

stfra-odoo committed:
[FIX] estate: correct the state modification of a property.

The state of property didn't change when a new offer was made and it still had no offer.

Now property state goes from 'New' to 'Offer received' when an offer is made.

stfra-odoo committed:
[IMP] estate_account: create a new model Estate accounting to link invoicing module to estate module.

Estate Accounting module depends on Estate and Account module to directly create an invoice the customer when a property is sold. The invoice has 3 lines:
- property price
- 6% fee
- Addimistrative fee

stfra-odoo committed:
[IMP] estate: add a basic kanban view to estate properties.

Kanban view cards are grouped by type and  display the name of the property the expected price, the best price (when an offer is recieved) and the selling price (when an offer is accepted).

ushyme commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,116 @@
+            self.garden_orientation = 'north'
+        else:
+            self.garden_area = 0
+            self.garden_orientation = ''
```
Use `False` (or `None`) to clear a selection field. Using an empty string `''` works but isn't standard practice in Odoo.
```suggestion
            self.garden_orientation = False
```

ushyme commented on estate/models/res_users.py:
```diff
@@ -0,0 +1,8 @@
+
+class ResUsers(models.Model):
+    _inherit = "res.users"
+    _name = "res.users"
```
```suggestion
```
Not needed when you are inheriting a standard Odoo model using `_inherit`.

ushyme commented on estate_account/security/ir.model.access.csv:
You can delete this file. You have already defined these exact access rights in the `estate` module. `estate_account` depends on estate, so it inherits these rights automatically. You only need this file if you are adding new models or modifying rights.

ushyme commented on estate_account/models/estate_property.py:
```diff
@@ -0,0 +1,47 @@
+                    }),
+                ]
+            }
+        ]
```
```suggestion
        invoice_vals_list = []
        for record in self:
            invoice_vals_list.append({
                'partner_id': record.property_buyer_id.id,
                'move_type': 'out_invoice',
                'invoice_line_ids': [
                    Command.create({
                        'name': record.name,
                        'quantity': 1,
                        'price_unit': record.selling_price
                    }),
                    Command.create({
                        'name': "6%",
                        'quantity': 1,
                        'price_unit': record.selling_price * 0.06
                    }),
                    Command.create({
                        'name': "Administrative fees",
                        'quantity': 1,
                        'price_unit': 100.0
                    }),
                ]
            })
```
The current implementation will raise an error if `action_sold` is called on multiple records. You should build `invoice_vals_list` for all records.

ushyme reviewed (commented):
Almost there! Also, all your XML files are missing a newline at the end of the file.
